src/scipp/core/binning.py

```diff
@@ -0,0 +1,494 @@
+                erase = erase.union(coord.dims)
+    return [dim for dim in erase if dim not in dims]
+
+
+def hist(x: Union[_cpp.DataArray, _cpp.Dataset],
+         arg_dict: Optional[Dict[str, Union[int, Variable]]] = None,
+         /,
+         **kwargs: Union[int, Variable]) -> Union[_cpp.DataArray, _cpp.Dataset]:
+    """Compute a histogram.
+
+    Bin edges can be specified in three ways: (1) When an integer is provided,
+    a 'linspace' with this requested number of bins is created, based on the min
+    and max of the corresponding coordinate. (2) A scalar scipp variable (a value
+    with a unit) is interpreted as a target bin with, and an 'arange' covering the
+    min and max of the corresponding coordinate is create. (3) A custom coordinate,
+    given as a scipp variable with compatible unit. Typically this should have a single
+    dimension matching the target dimension.
+
+    When histogramming a dimension with an existing dimension-coord, the binning for
+    the dimension is modified, i.e., the input and the output will have the same
+    dimension labels.
+
+    When histogramming by non-dimension-coords, the output will have new dimensions
+    given by the names of these coordinates. These new dimensions replace the
+    dimensions the input coordinates depend on.
+
+    Parameters
+    ----------
+    x:
+        Input data.
+    arg_dict:
+        Dictionary mapping dimension labels to binning parameters.
+    **kwargs:
+        Mapping of dimension label to correspoding binning parameters.
+
+    Returns
+    -------
+    :
+        Histogrammed data.
+
+    See Also
+    --------
+    scipp.bin:
+        Creating binned data by binning instead of summing all contributions.
+    scipp.binning.make_histogrammed:
+        Lower level function for histogramming that does not automatically
+        replace/erase dimensions.
+
+    Examples
+    --------
+
+    Histogram a table by one of its coord columns, specifying (1) number of bins, (2)
+    bin width, or (3) actual binning:
+
+      >>> table = sc.data.table_xyz(100)
+      >>> da = table.hist(x=2)
+      >>> da = table.hist(x=sc.scalar(0.2, unit='m'))
+      >>> da = table.hist(x=sc.linspace('x', 0.2, 0.8, num=10, unit='m'))
+
+    Histogram a table by two of its coord columns:
+
+      >>> table = sc.data.table_xyz(100)
+      >>> da = table.hist(x=4, y=6)
+
+    Histogram binned data, using existing bins:
+
+      >>> binned = sc.data.binned_x(nevent=100, nbin=10)
+      >>> da = binned.hist()
+
+    Histogram binned data, using new bins along existing dimension:
+
+      >>> binned = sc.data.binned_x(nevent=100, nbin=10)
+      >>> da = binned.hist(x=20)
+
+    Histogram binned data along an additional dimension:
+
+      >>> binned = sc.data.binned_x(nevent=100, nbin=10)
+      >>> da = binned.hist(y=5)
+    """
+    edges = _make_edges(x, arg_dict, kwargs)
+    erase = _find_replaced_dims(x, edges)
+    if len(edges) == 0:
+        return x.bins.sum()
```

Comment:
Do we need to safeguard against calling `.hist()` on data that is not binned? I guess currently we would get the error that `None has no attribute 'sum'` (or something like that)? Basically, do we need a better error message?

Reply:
It is perfectly fine to call `hist` on non-binned data (a 1-D table).

Reply:
Yes, but don't you then need to supply edges? i.e. the list of arguments cannot be empty in that case. If not you woudn't know how to histogram?

Reply:
Good point, fixed now!